exporter/exporterhelper/batch_sender.go

```diff
@@ -0,0 +1,327 @@
+		stopped:      &atomic.Bool{},
+		resetTimerCh: make(chan struct{}),
+	}
+
+	for _, op := range opts {
+		op(bs)
+	}
+	return bs
+}
+
+func (bs *batchSender) Start(_ context.Context, _ component.Host) error {
+	timer := time.NewTimer(bs.mergeCfg.Timeout)
+	go func() {
+		for {
+			select {
+			case <-bs.shutdownCh:
+				bs.mu.Lock()
+				if bs.activeBatch.request != nil {
+					bs.exportActiveBatch()
+				}
+				bs.mu.Unlock()
+				return
+			case <-timer.C:
+				bs.mu.Lock()
+				if bs.activeBatch.request != nil {
+					bs.exportActiveBatch()
+				}
+				bs.mu.Unlock()
+				timer.Reset(bs.mergeCfg.Timeout)
+			case <-bs.resetTimerCh:
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(bs.mergeCfg.Timeout)
+			}
+		}
+	}()
+
+	return nil
+}
+
+type batch struct {
+	ctx     context.Context
+	request Request
+	done    chan struct{}
+	err     error
+}
+
+func newEmptyBatch() *batch {
+	return &batch{
+		ctx:  context.Background(),
+		done: make(chan struct{}),
+	}
+}
+
+// exportActiveBatch exports the active batch and creates a new one.
+// Caller must hold the lock.
+func (bs *batchSender) exportActiveBatch() {
+	go func(b *batch) {
+		b.err = b.request.Export(b.ctx)
+		close(b.done)
+	}(bs.activeBatch)
+	bs.activeBatch = newEmptyBatch()
+}
+
+// isActiveBatchReady returns true if the active batch is ready to be exported.
+// The batch is ready if it has reached the minimum size or the concurrency limit is reached.
+// Caller must hold the lock.
+func (bs *batchSender) isActiveBatchReady() bool {
+	return bs.activeBatch.request.ItemsCount() >= bs.mergeCfg.MinSizeItems || bs.activeRequests.Load() >= bs.concurrencyLimit
+}
+
+func (bs *batchSender) send(ctx context.Context, req Request) error {
+	// Stopped batch sender should act as pass-through to allow the queue to be drained.
+	if bs.stopped.Load() {
+		return bs.nextSender.send(ctx, req)
+	}
+
+	bs.activeRequests.Add(1)
+	defer bs.activeRequests.Add(^uint64(0))
+
+	if bs.mergeSplitFunc != nil {
+		return bs.sendMergeSplitBatch(ctx, req)
+	}
+	return bs.sendMergeBatch(ctx, req)
+}
+
+func (bs *batchSender) sendMergeSplitBatch(ctx context.Context, req Request) error {
+	bs.mu.Lock()
+
+	reqs, err := bs.mergeSplitFunc(ctx, bs.activeBatch.request, req, bs.splitCfg.MaxSizeItems)
+	if err != nil || len(reqs) == 0 {
+		bs.mu.Unlock()
+		return err
+	}
+	if len(reqs) == 1 || bs.activeBatch.request != nil {
+		bs.activeBatch.request = reqs[0]
+		bs.activeBatch.ctx = ctx
+		batch := bs.activeBatch
+		var sent bool
+		if bs.isActiveBatchReady() || len(reqs) > 1 {
+			bs.exportActiveBatch()
+			sent = true
+		}
+		bs.mu.Unlock()
+		if sent {
+			bs.resetTimerCh <- struct{}{}
+		}
+		<-batch.done
+		if batch.err != nil {
+			return batch.err
+		}
+		reqs = reqs[1:]
+	} else {
+		bs.mu.Unlock()
+	}
+
+	// Intentionally do not put the last request in the active batch to not block it.
+	// TODO: Consider including the partial request in the error to avoid double publishing.
```

Comment:
If I'm reading the code and TODO correctly -- it looks like the batch `Export()` error may or may not be returned to the caller depending on whether the export is triggered by a timer or by a full batch.  When one caller sees the error, it's because a batch that (may have?) included its own data failed synchronously while trying to send. The error handling looks inaccurate.

I'm looking to replicate the functionality in https://github.com/open-telemetry/otel-arrow/blob/main/collector/processor/concurrentbatchprocessor/README.md, which has each caller block until the data they entered into one or more batches has been decisively exported or not. Each caller ends up with different partials depending on whether the data they submitted lands in one or more batches. If the data they submitted ends in exactly one batch, it will not get a partial error in that case, it will get a total error describing the uniform result. 

The solution I'm referring to blocks each caller until each caller's data has been processed and returns one of three outcomes to each caller: total error, partial error, and no error. It looks like the solution here may or may not block the caller until their data is processed.

The reason this matters is that I want to have accurate success/failure counts from the producer perspective. If the batch processor does not return accurate error information to the SDKs that produced the data, then the SDK-level metrics become not very useful. It's not only to avoid double publishing, it's to improve SDK-level metrics accuracy.

Reply:
The current solution always blocks all the callers until the batch is complete or timed out. 

This particular comment is about splitting behavior. If one of the batch splits fails to be sent, we return an error as the whole batch failed. We should return a partial error, letting the caller know what portion failed. I can take care of this TODO in this PR. Addressing it should cover all your needs

Reply:
BTW the line above is not related to the TODO. It just means that we send the remainder of a split right away, even if the size of it is smaller than the minimum threshold. It's needed to avoid stacking blocked requests on top of each other if they are not part of one batch.